core/storage/statistics/gae_models.py

```diff
@@ -737,3 +845,247 @@ def _get_entity_id(cls, exploration_id, exploration_version, state_name,
+            large_answer_bucket_count)
+
+    @classmethod
+    def _finish_migrating_answer(
+            cls, state_answer_log_model_item_id, exploration_version):
+        model = cls.get(state_answer_log_model_item_id)
+        # Avoid an unnecessary put if the given version is already recorded
+        if exploration_version not in model.exploration_versions:
+            model.exploration_versions.append(exploration_version)
+            model.put()
+
+    @classmethod
+    def finish_migrating_answer(
+            cls, state_answer_log_model_item_id, exploration_version):
+        transaction_services.run_in_transaction(
+            cls._finish_migrating_answer, state_answer_log_model_item_id,
+            exploration_version)
+
+    @classmethod
+    def _finish_migration_answer_bucket(
+            cls, state_answer_log_model_item_id, large_answer_bucket_id):
+        model = cls.get(state_answer_log_model_item_id)
+        if large_answer_bucket_id:
+            finished_large_answer_bucket_ids = set(
+                model.finished_large_answer_bucket_ids)
+            finished_large_answer_bucket_ids.add(large_answer_bucket_id)
+            expected_bucket_count = model.expected_large_answer_bucket_count
+            model.finished_migration = (
+                len(finished_large_answer_bucket_ids) >= expected_bucket_count)
+            model.finished_large_answer_bucket_ids = list(
+                finished_large_answer_bucket_ids)
+        else:
+            model.finished_migration = True
+        model.put()
+
+    @classmethod
+    def finish_migration_answer_bucket(
+            cls, state_answer_log_model_item_id, large_answer_bucket_id):
+        transaction_services.run_in_transaction(
+            cls._finish_migration_answer_bucket, state_answer_log_model_item_id,
+            large_answer_bucket_id)
+
+    @classmethod
+    def has_started_being_migrated(
+            cls, state_answer_log_model_item_id, large_answer_bucket_id=None):
+        model = cls.get(state_answer_log_model_item_id, strict=False)
+        return model is not None and (
+            not large_answer_bucket_id or large_answer_bucket_id
+            in model.started_large_answer_bucket_ids)
+
+    @classmethod
+    def validate_answers_are_migrated(cls, state_rule_answer_log_model):
+        migrated_answer_model = MigratedAnswerModel.get(
+            state_rule_answer_log_model.id, strict=False)
+        if not migrated_answer_model:
+            raise utils.ValidationError(
+                u'Answers not migrated: %s' % state_rule_answer_log_model.id)
+        state_answer_models_list = []
+
+        # A version of -1 is a special sentinel value to silence the
+        # validation on this answer bucket. It typically represents an
+        # answer which cannot be migrated for a known reason. These answers
+        # should not be validated, since they were never migrated.
+        if migrated_answer_model.exploration_versions == [-1]:
```

Comment:
Worth using a named constant? "-1" is quite generic.

Also, does one non-migrated answer spoil the whole batch? Would it be better to instead keep a count of unmigratable answers?